state/mapstate/map_state.go

```diff
@@ -92,22 +97,70 @@ func (st *MapState) Snapshot(w io.Writer) error {
 	return enc.Encode(st)
 }
 
-// Restore takes a reader and restores a snapshot. It should migrate
-// the format if it is not compatible with the current version.
-func (st *MapState) Restore(r io.Reader) error {
-	snap, err := ioutil.ReadAll(r)
-	if err != nil {
-		return err
+// Restore restores a snapshot from the state's internal bytes. It should
+// migrate the format if it is not compatible with the current version.
+func (st *MapState) Restore() error {
+	if st.Version == Version {
+		return nil
 	}
-	var vonly struct{ Version int }
-	err = json.Unmarshal(snap, &vonly)
+	err := st.migrateFrom(st.Version, st.toRestore)
 	if err != nil {
 		return err
 	}
-	if vonly.Version == Version {
-		// we are good
-		err := json.Unmarshal(snap, st)
+	st.Version = Version
+	st.toRestore = nil
+	return nil
+}
+
+// GetVersion returns the current version of this state object.
+// It is not necessarily up to date
+func (st *MapState) GetVersion() int {
+	return st.Version
+}
+
+// GetLatestVersion returns the most up to date version of state
+func (st *MapState) GetLatestVersion() int {
+	return Version
+}
+
+// Marshal encodes the state using msgpack
+func (st *MapState) Marshal() ([]byte, error) {
+	logger.Debugf("Marshal-- Marshalling state of version %d", st.Version)
+	buf := new(bytes.Buffer)
+	enc := msgpack.Multicodec(msgpack.DefaultMsgpackHandle()).Encoder(buf)
+	if err := enc.Encode(st); err != nil {
+		return nil, err
+	}
+	// First byte indicates the version (probably should make this a varint
+	// if we stick to this encoding)
+	vCodec := make([]byte, 1)
+	vCodec[0] = byte(st.Version)
+	ret := append(vCodec, buf.Bytes()...)
+	logger.Debugf("Marshal-- The final marshaled bytes: %x", ret)
+	return ret, nil
+}
+
+// Unmarshal decodes the state using msgpack.  It first decodes just
+// the version number.  If this is not the current version the bytes
+// are stored within the state's internal reader, which can be migrated
+// to the current version in a later call to restore.  Note: Out of date
+// version is not an error
+func (st *MapState) Unmarshal(bs []byte) error {
+	// Check version byte
+	logger.Debugf("The incoming bytes to unmarshal: %x", bs)
+	v := int(bs[0])
+	logger.Debugf("The interpreted version: %d", v)
+	if v != Version { // snapshot is out of date
+		st.Version = v
+		st.toRestore = bs[1:]
```

Comment:
I find it a bit indirect to set `.toRestore` here. Is it possible to just return an error and then call:

`st.Restore(bs []byte)`. Or `st.Migrate(bs []byte)`

The "indirection" that bothers me is that it is not clear that Restore can only be called after a failed Unmarshal() attempt.

Reply:
Getting rid of this with the new approach suggested a few comments up 🎉